File: leetcode_python/lib/problems/p0019_remove_nth_node_from_end_of_list/remove_nth_node_from_end_of_list.py
from typing import List, Optional

from lib.helpers.int_list_node import ListNode


class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        if head == None:
            return None

        stack = []
        node = head
        while node:
            stack.append(node)
            node = node.next

        node = None
        while n > 0:
            node = stack.pop()
            n -= 1

        if len(stack) == 0:
            return node.next

        topStack = stack[len(stack) - 1]
        topStack.next = node.next

        return head

File: leetcode_js/src/removeNthNodeFromEndOfList.js
const { ListNode } = require('../tests/helpers/listNode');

const removeNthFromEnd = function (head, n) {
  if (head === null) {
    return null;
  }

  const dummy = new ListNode(null);
  dummy.next = head;

  const stack = [];
  let node = dummy;
  while (node) {
    stack.push(node);
    node = node.next;
  }

  for (let i = 0; i < n; i++) {
    node = stack.pop();
  }

  const nextNode = node.next;
  const prevNode = stack[stack.length - 1];

  node.next = null;
  prevNode.next = nextNode;

  return dummy.next;
};

module.exports = { removeNthFromEnd };
